pulsar-zookeeper-utils/src/main/java/com/yahoo/pulsar/zookeeper/ZooKeeperCache.java

```diff
@@ -285,32 +286,57 @@ public Boolean call() throws Exception {
         CompletableFuture<Optional<Entry<T, Stat>>> future = new CompletableFuture<>();
         dataCache.get(path, (p, executor) -> {
             // Return a future for the z-node to be fetched from ZK
+            final CompletableFuture<Boolean> existsFuture = new CompletableFuture<>();
             CompletableFuture<Entry<Object, Stat>> zkFuture = new CompletableFuture<>();
 
             // Broker doesn't restart on global-zk session lost: so handling unexpected exception
             try {
-                this.zkSession.get().getData(path, watcher, (rc, path1, ctx, content, stat) -> {
-                    Executor exec = scheduledExecutor != null ? scheduledExecutor : executor;
+                // Check for existence first, then get data: this allows us to put the watch on unconditionally.
+                zkSession.get().exists(path, watcher, (rc, path1, ctx, stat) -> {
```

Comment:
I actually saw the speed issue in earlier editions of the code where I did that. It also adds a bit of complexity to the code, because to do it properly we need to make sure that if the node does exist after calling `exists`, we should try to get the data again, to ensure that we have the most recent version of the data.